### PR TITLE
server: inject regex engine before validation

### LINKED ISSUE #24022: Config validation command fails when deprecated regex engine property not present
*Title*: Config validation command fails when deprecated regex engine property not present 

*Description*:
Since 1.23, regex engine is no longer needed per https://github.com/envoyproxy/envoy/pull/21633
But removing it will fail validation command, potential related to https://github.com/envoyproxy/envoy/pull/21972
Note it can run well in standard mode.

```
regex_rewrite:
  pattern:
    google_re2: {} # comment this out would result in `envoy --mode validation` crash
    regex: "some"
```

*Repro steps*:
Use below config.yaml
```
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 9901 }

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address: { address: 127.0.0.1, port_value: 10000 }
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route:
                  regex_rewrite:
                    pattern:
                      # google_re2: {}
                      regex: "some"
                    substitution: ""
                  cluster: some_service
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
  clusters:
  - name: some_service
    connect_timeout: 0.25s
    type: STATIC
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: some_service
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 127.0.0.1
                port_value: 1234
```

envoy version 1.24.0
Run
```
envoy --mode validate -c config.yaml
```
output:
```
[2022-11-16 08:18:30.377][36][info][main] [source/server/server.cc:808] runtime: {}
[2022-11-16 08:18:30.377][36][info][config] [source/server/configuration_impl.cc:131] loading tracing configuration
[2022-11-16 08:18:30.377][36][info][config] [source/server/configuration_impl.cc:91] loading 0 static secret(s)
[2022-11-16 08:18:30.377][36][info][config] [source/server/configuration_impl.cc:97] loading 1 cluster(s)
[2022-11-16 08:18:30.378][36][info][config] [source/server/configuration_impl.cc:101] loading 1 listener(s)
[2022-11-16 08:18:30.379][36][critical][assert] [./source/common/singleton/threadsafe_singleton.h:56] assert failure: loader_ != nullptr. Details: InjectableSingleton used prior to initialization
[2022-11-16 08:18:30.379][36][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x24
[2022-11-16 08:18:30.379][36][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2022-11-16 08:18:30.379][36][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 15baf56003f33a07e0ab44f82f75a660040db438/1.24.0/Clean/RELEASE/BoringSSL
[2022-11-16 08:18:30.380][36][critical][backtrace] [./source/server/backtrace.h:96] #0: __restore_rt [0x7fd8ae40e420]
[2022-11-16 08:18:30.380][36][critical][backtrace] [./source/server/backtrace.h:98] #1: [0x55fa463ed761]
[2022-11-16 08:18:30.380][36][critical][backtrace] [./source/server/backtrace.h:98] #2: [0x55fa463f4ae5]
...
[2022-11-16 08:18:30.384][36][critical][backtrace] [./source/server/backtrace.h:98] #33: [0x55fa4486c11c]
[2022-11-16 08:18:30.384][36][critical][backtrace] [./source/server/backtrace.h:96] #34: __libc_start_main [0x7fd8ae22c083]
```

If uncomment `google_re2: {}`, output is:
```
[2022-11-16 08:23:55.469][46][info][main] [source/server/server.cc:808] runtime: {}
[2022-11-16 08:23:55.469][46][info][config] [source/server/configuration_impl.cc:131] loading tracing configuration
[2022-11-16 08:23:55.469][46][info][config] [source/server/configuration_impl.cc:91] loading 0 static secret(s)
[2022-11-16 08:23:55.469][46][info][config] [source/server/configuration_impl.cc:97] loading 1 cluster(s)
[2022-11-16 08:23:55.470][46][info][config] [source/server/configuration_impl.cc:101] loading 1 listener(s)
[2022-11-16 08:23:55.473][46][warning][misc] [source/common/protobuf/message_validator_impl.cc:21] Deprecated field: type envoy.type.matcher.v3.RegexMatcher Using deprecated option 'envoy.type.matcher.v3.RegexMatcher.google_re2' from file regex.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/version_history/version_history for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override.
[2022-11-16 08:23:55.473][46][warning][misc] [source/common/protobuf/message_validator_impl.cc:21] Deprecated field: type envoy.type.matcher.v3.RegexMatcher Using deprecated option 'envoy.type.matcher.v3.RegexMatcher.google_re2' from file regex.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/version_history/version_history for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override.
[2022-11-16 08:23:55.475][46][info][config] [source/server/configuration_impl.cc:113] loading stats configuration
```

*Admin and Stats Output*:

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/main/bazel#stack-trace-symbol-resolution).

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: Xie Zhihao <zhihao.xie@intel.com>

Commit Message: server: inject regex engine before validation
Additional Description:
Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: Added
Platform Specific Features: N/A
Fixes #24022

### CODE REVIEW

File: source/server/config_validation/server.cc [hunk 1]
```diff
@@ -82,6 +82,22 @@ void ValidationInstance::initialize(const Options& options,
...
+        Config::Utility::getAndCheckFactory<Regex::EngineFactory>(default_regex_engine);
+    auto config = Config::Utility::translateAnyToFactoryConfig(
+        default_regex_engine.typed_config(), messageValidationContext().staticValidationVisitor(),
+        factory);
+    regex_engine = factory.createEngine(*config, serverFactoryContext());
+  } else {
+    regex_engine = std::make_shared<Regex::GoogleReEngine>();
+  }
+  Regex::EngineSingleton::clear();
+  Regex::EngineSingleton::initialize(regex_engine.get());
```
Comment (zuercher) [inline source/server/config_validation/server.cc:99]: What do you think of moving this block to a helper function in `source/server`? (For example like `createContextManager` in ssl_context_manager.h? Except that in Server::Instance the result is assigned to a field I think it's identical.

**Review by zuercher — COMMENTED** [review]
Thanks! That looks good to me, with one comment.

File: changelogs/current.yaml [hunk 2]
```diff
@@ -47,6 +47,9 @@ bug_fixes:
 - area: http
   change: |
     fixed a bug where Utility::PercentEncoding::encode() encodes some characters incorrectly because it was treating the value as negative.
+- area: validation
+  change: |
+    fixed a crash wihch could happen when optional ``engine_type`` is not provided in regex.
```
Comment (karataliu) [inline changelogs/current.yaml:52]: typo-> which

Comment (zhxie) [reply changelogs/current.yaml:52]: Thanks, fixed.

File: source/server/config_validation/server.cc (hunk 1 above)
Comment (zhxie) [reply source/server/config_validation/server.cc]: Thanks, fixed.

**Comment by zuercher** [comment]
Maybe @adisuissa could take a look for a second set of eyes?

**Review by zuercher — APPROVED** [review]
Looks good to me. Bonus points if you wanted to add a test, but we don't seem to have a good way of testing all the things that need to be initialized in both the real server and the validation instance.

**Review by adisuissa — APPROVED** [review]
LGTM.
nit: I suggest that the PR title will say config_validation or server_config_validation, as it is not really impacts the server but the config_validation mode.

**Pull request merged**